Add Cypress PSoC 6 targets

vmedcy opened the pull request:
### Description

Initial support for Cypress boards at targets/TARGET_Cypress/TARGET_PSOC6.

Added pioneer kits:
* CY8CKIT-062-4343W
* CY8CKIT-062-BLE
* CY8CKIT-WIFI-BT

Added prototyping boards:
* CY8CPROTO-062-4343W

Added connectivity modules:
* CY8CMOD-062-4343W

The pull request is split into multiple commits:

1. Move existing FUTURE_SEQUANA port to targets/TARGET_Cypress/TARGET_PSOC6_FUTURE

2. Add Cypress PSoC 6 HAL, latest PDL, support for 4 board targets and 4343W connectivity module.

3. Add mbedtls target providing alternate implementation for PSoC 6 crypto hardware.

4. Add BLE HCI drivers for CYW43012 and CYW4343W.

5. Add WICED connectivity drivers compiled for all supported boards and toolchains.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Related pull requests

PSOC6.py updates: https://github.com/ARMmbed/mbed-os/pull/9466 https://github.com/ARMmbed/mbed-os/pull/9509

platform database update: https://github.com/ARMmbed/mbed-os-tools/pull/110

### Reviewers

@0xc0170 @lrusinowicz @orenc17

vmedcy commented:
Test suite executed on CY8CPROTO_062_4343W still reports several failures, this is investigated:

mbed test -m CY8CPROTO_062_4343W -t GCC_ARM
...
mbedgt: test case results: 1 FAIL / 15 SKIPPED / 601 OK / 3 ERROR

[CY8CPROTO_062_4343W_GCC_ARM_20190123.log](https://github.com/ARMmbed/mbed-os/files/2790540/CY8CPROTO_062_4343W_GCC_ARM_20190123.log)

ciarmcom commented:
@vmedcy, thank you for your changes.
@0xc0170 @orenc17 @ARMmbed/mbed-os-crypto @ARMmbed/mbed-os-pan @ARMmbed/mbed-os-maintainers please review.

mikisch81 commented:
@ARMmbed/mbed-os-psa

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_062_WIFI_BT/GeneratedSource/cycfg_notices.h:
```diff
@@ -0,0 +1,21 @@
+* 
+********************************************************************************
+* Copyright (c) 2017-2018 Cypress Semiconductor.  All rights reserved.
+* You may use this file only in accordance with the license, terms, conditions, 
```
What license for these files? It's missing in this header (it references user agreement, cant spot any in the wifi bt folder). 
There were similar files already integrated, they were released under Apache 2.0

vmedcy replied:
The low-level PSoC initialization code in Cypress port is generated by ModusToolbox MCU Configurators (unlike FUTURE_SEQUANA port which used PSoC Creator tool). The configurators are being updated to generate code with Apache license header.
I will regenerate the code with latest configurators to apply the new license. No Cypress EULA included in mbed port, this is intended.

vmedcy replied:
@0xc0170: this is addressed now, all BSP code in GeneratedSource is now Apache 2.0.

vmedcy replied:
@0xc0170: together with the latest rebase, I included the CapSense middleware: 
5252d05. The code is supplied with the standard Cypress EULA, it is not Apache 2.0 licensed. Per the guidelines, the license file is added to the psoc6mw/capsense folder.

lrusinowicz reviewed (changes requested):
I'm not convinced the way it was done, i.e. copying all the files for the basic device support is the best option. This will definitely complicate maintenance in the future.
I understand you want to use newer PDL version, but I see no reason why Sequana would not work with it, if the CY8CKIT_062_BLE does. That's basically the same chip. I'm switching between these boards for testing purposes all the time. Just update PDL and be done with it.

lrusinowicz reviewed (commented):
I see that you have removed the "chip" layer from the vendor->family->chip->board hierarchy that I've originally implemented. This greatly increases amount of copy-pasting when one want's to add a new board, as it can be seen even in your commit.
What's motivation behind this?

vmedcy commented:
@lrusinowicz
> I understand you want to use newer PDL version, but I see no reason why Sequana would not work with it, if the CY8CKIT_062_BLE does. That's basically the same chip. I'm switching between these boards for testing purposes all the time. Just update PDL and be done with it.

The desire is to limit the scope of this PR to only add Cypress kits support, but don't break existing SEQUANA boards. Cypress cannot test the SEQUANA targets, due to the limited board availability and different initialization sequence in CM0+ prebuilt images; also the work on _M0 and _PSA targets have not been started.
I propose you to migrate the Sequana port over time to use the Cypress maintained TARGET_PSOC6, including the latest PDL and GeneratedSource generated by ModusToolbox. But this cannot be done all at once.

vmedcy commented:
>  What's motivation behind this?

There is a valid reason to provide separate device support files per each BSP target (PeripheralNames.h, PeripheralPins.c, PinNames.h, system_psoc6.h/.c, linker scripts, startup assembly and prebuilt CM0+ images). Different Cypress boards have different PSoC 6 MPNs with different FLASH/SRAM layouts (CY8C6XX6, CY8C6XX7, CY8C6XXA) and different hardware dies (PSoC6ABLE2, PSoC6A2M). Each memory layout requires separate set of linker scripts. Each hardware die requires separate startup assembly and prebuilt M0 image. Having this stuff maintained at the family level and selected by extra labels in targets.json would introduce very complex directory, up to the point when the Windows MAX_PATH limit is hit: targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CXX7/TARGET_PSOC6ABLE2/TOOLCHAIN_GCC_ARM/cy8c6xx7_cm4_dual.ld and so on.
Another important point is the selection of the prebuilt CM0+ image: different images might consume different amount of RAM/FLASH (deepsleep, BLE, security, ..). For larger images, modification of the linker scripts and CY_CORTEX_M4_APPL_ADDR in system_psoc6.h is required at the moment (this can be reduced to the absolute minimum in the future, by json parametrization).
The proposed layout can handle this easily. Also it is super intuitive for customers who go to production with alternate MPNs and custom devboards how to create their own custom target - just copy the existing one and adjust as needed. Customers who target connectivity module (CY8CMOD_062_4343W) can copy existing TARGET_CY8CPROTO_062_4343W sub-target to only adjust PinNames.h to match their board.

Everything that can be shared across boards (CMSIS device headers and peripheral drivers) is already in psoc6pdl directory - this is where most development/maintenance happens.

cmonr commented:
@vmedcy Ooph, this is quite the PR! 

While the relevant reviews take a look at this PR, a couple of things. 
1) 👍 on the commit comments, as well as on their topic deliniations!
Even though this PR is too big for GitHub to load all of the files at once (it paginated three times for me) we should be able to use the commits to deliniate the review topics. 
2) https://github.com/ARMmbed/mbed-os/pull/9481#issuecomment-457090712
 - Thanks for provinding the GCC test one of the targets. Since this is only one target, would you mind providing the test results for the other targets being added with this PR?
 - We also need test results for the ARM compiler (ARMc5u6) and IAR (7.8).

cmonr commented:
I thought I forgot something.

3. The GCC test, it needs to be completely passing. I noticed there were a couple of failures and timeouts that should be corrected.

cmonr commented:
While I'm here, it looks like GitHub noticed something else.

4. This PR will need to be rebased, to resolve this conflicting file:
` targets/TARGET_Cypress/TARGET_PSOC6/TARGET_FUTURE_SEQUANA_PSA/prebuilt/README.md `

Note: When done correctly, the PR will not have any merge commits in the history. (Sometimes we find that people get merges and rebases confused)

vmedcy commented:
@cmonr:
>     1. This PR will need to be rebased, to resolve this conflicting file:
>        `targets/TARGET_Cypress/TARGET_PSOC6/TARGET_FUTURE_SEQUANA_PSA/prebuilt/README.md`
> 
This is done, rebased against the latest master, absolutely no problem with that (force pushed the PR branch).
I also included the latest HAL updates and GeneratedSource license fixes.

>     The GCC test, it needs to be completely passing. I noticed there were a couple of failures and timeouts that should be corrected.

This is fully understood. The team still works on getting the full test suite passing.

>     We also need test results for the ARM compiler (ARMc5u6) and IAR (7.8).

Attached greentea test results on CY8CKIT_062_WIFI_BT with GCC_ARM 6.3.1, ARM Compiler 5.06 update 6, and IAR Compiler V8.20.
[CY8CKIT_062_WIFI_BT-IAR-20190124.log](https://github.com/ARMmbed/mbed-os/files/2795061/CY8CKIT_062_WIFI_BT-IAR-20190124.log)
[CY8CKIT_062_WIFI_BT-ARM-20190124.log](https://github.com/ARMmbed/mbed-os/files/2795062/CY8CKIT_062_WIFI_BT-ARM-20190124.log)
[CY8CKIT_062_WIFI_BT-GCC_ARM-20190124.log](https://github.com/ARMmbed/mbed-os/files/2795063/CY8CKIT_062_WIFI_BT-GCC_ARM-20190124.log)

Test reports for other PSoC 6 targets will be attached later.

There are still several failures reported (tests-mbed_hal-sleep_manager, tests-mbed_hal-sleep, tests-mbed_drivers-lp_timeout, features-feature_ble-targets-target_cordio-tests-cordio_hci-transport):

mbedgt: test case results: 3 FAIL / 1 SKIPPED / 615 OK / 1 ERROR

Supporting IAR Compiler V7.8 requires extra work: WICED binary drivers were prebuilt with IAR 8, not compatible with IAR 7. Need to investigate possibility to support both IAR 7 and 8. I believe Mbed-os moves to IAR 8 soon, per https://github.com/ARMmbed/mbed-os-5-docs/pull/885. Can you suggest if any IAR 7 related effort is justified at this point of time?

pan- commented on features/FEATURE_BLE/targets/TARGET_Cypress/TARGET_CYW43012/HCIDriver.cpp:
```diff
@@ -92,16 +75,30 @@ class HCIDriver : public cordio::CordioHCIDriver {
-        output_mode(bt_device_wake_name, 0);
-        output_mode(bt_power_name, 1);
-
+        //output_mode(bt_host_wake_name, 1);
```
Could you comment that change ? Is there any reason to keep that code commented ?

vmedcy replied:
Removed commented code with the latest update: https://github.com/ARMmbed/mbed-os/pull/9481/commits/8ca1720e7bbc60688ab49da6bd9ebe1735e64557.

Regarding duplication of HCI implementation: the idea is to design this for extensibility from the beginning.
The pin assignment for CYW43012 and CYW4343X HCI drivers just happens to match for the 4 currently supported boards. In the future, as we add more boards, it may diverge and require  different config.
Since CYW43012 has embedded stack mode, as we may add this feature support in future, it may also make the HCIDriver differ from 4343W.

vmedcy replied:
@pan-: the team put more thought on HCI driver design. The driver duplication is now addressed by https://github.com/ARMmbed/mbed-os/pull/9481/commits/6a361837323a4058d68a6fea519cb16e1e6a54dc: there is a single driver for CYW4343W and CYW43012 in features/FEATURE_BLE/targets/TARGET_Cypress/TARGET_CYW43XXX. Differences in the implementation are to be handled by constructor argument and json parametrization.
Can you resolve this conversation?

pan- reviewed (changes requested):
`TARGET_CYW43012/HCIDriver.cpp` and `TARGET_CYW4343X/HCIDriver.cpp` are identical. I believe it would be better for maintenance purposes to have a single driver rather than duplicate code.

lrusinowicz commented:
@vmedcy 
> 
> The desire is to limit the scope of this PR to only add Cypress kits support, but don't break existing SEQUANA boards. Cypress cannot test the SEQUANA targets, due to the limited board availability and different initialization sequence in CM0+ prebuilt images; also the work on _M0 and _PSA targets have not been started.
> I propose you to migrate the Sequana port over time to use the Cypress maintained TARGET_PSOC6, including the latest PDL and GeneratedSource generated by ModusToolbox. But this cannot be done all at once.

So more work for us, because you want to avoid some, good.
I don't expect you to test Sequana, but OTOH, Mbed tests for Sequana could be run on CY8CKIT_062_BLE with just proper name mocking (mbedls -mock option), assuming it's native tests do work.
You could have both PDLs at the same level and select a proper one using json file if you'd move PDLs to directories prefixed with TARGET_. This way no copy of the BSP would be needed.

You are talking about the Windows path length limit elsewhere, but you are lengthening it for us.

pan- commented:
@vmedcy Few more questions: 
- I want to validate the port on a `CY8CKIT_062_BLE`. Do I need to flash a special firmware ? My board is not recognised by mbedls and the ID listed in https://github.com/ARMmbed/mbed-os-tools/pull/110 doesn't match what is returned by the board: 

```
'IORegistryEntryName': 'Cypress KitProg2 (CMSIS-DAP, Mass Storage)',
'USB Serial Number': '10120C5303147400',
```

- Do I need https://github.com/ARMmbed/mbed-os/pull/9466 checked in to validate this PR ? 
- Why `BLE` is not a listed feature of `CY8CKIT_062_BLE` ?

vmedcy commented:
@lrusinowicz:
> You could have both PDLs at the same level and select a proper one using json file if you'd move PDLs to directories prefixed with TARGET_. This way no copy of the BSP would be needed.

Avoiding complex nested TARGET_ hierarchies is an explicit goal.  I agree this adds a little maintenance overhead (same linker scripts appear in different BSP target folders), but makes things much easier for end users, especially when creating custom BSPs.

vmedcy commented:
@pan-:
> I want to validate the port on a `CY8CKIT_062_BLE`

Great!

> Do I need to flash a special firmware?

Yes, CY8CKIT_062_BLE comes with KitProg2 FW, not compatible with mbedls. DAPLink firmware is under development, we flash nightly builds of KitProg3 FW during development/testing. Please suggest the convenient way for you to receive the compatible firmware (requires either special FW upload tool or Bootloader Host Tool supplied with PSoC Creator).

> Do I need #9466 checked in to validate this PR ?

Yes, as well as  #9509 I just submitted. To merge all 3 PRs:

git fetch origin pull/9481/head:add_cypress_kits
git fetch origin pull/9466/head:psoc6-target-hook
git fetch origin pull/9509/head:psoc6-daplink-hex
git checkout add_cypress_kits
git merge psoc6-target-hook
git merge psoc6-daplink-hex

> Why `BLE` is not a listed feature of `CY8CKIT_062_BLE` ?

BLE hardware is different in CY8CKIT_062_BLE (using on-chip BLESS instead of WICED connectivity module). It is not supported by the existing port. The FW implementation will be provided in consequent PRs.

c1728p9 commented:
I @vmedcy, thanks for the PR! I'm in the process of reviewing it, but am having trouble viewing diffs with the commit history in its current form. Some of the problems I'm running into when reviewing:

- The commit "Add Cypress PSoC 6 targets" says that some of the code is based on the FUTURE_SEQUANA. This commit only shows new files so it isn't easy to see what was changed for the new targets. Adding a commit before this which just copies over the resued files without any modifications would allow the the diff to be shown in git.
- The same commit, "Add Cypress PSoC 6 targets", has a large number of generated files in addition to changing files from the FUTURE_SEQUANA. If the code generation is split out into a seperate commit then the steps to generate this could be reviewed on its own.
- The commit "PSOC6: add Cypress target to mbedtls" has license updates and maybe bug fixes in addition the mbedtls support it adds. Each commit could be reviewed in more detail if this was broken into 3 commits - one for TLS, one for license updates and one for bug fixes.

Would it be possible to update the history of this PR to make it easier to review? I made these changes to the history in my fork of mbed-os so feel free to push that directly or with modifications if it is useful to you. The end result is identical, I just broke apart the history to make it easier to review. The updated history can be found in my [mbed-os repo on the branch pr9481_history](https://github.com/c1728p9/mbed-os/tree/pr9481_history). Commits of my branch:


1. Move FUTURE_SEQUANA port to TARGET_PSOC6_FUTURE (unchanged)
2. Reuse FUTURE_SEQUANA porting layer
    - I copied the files that I thought came from FUTURE_SEQUANA so I could see a diff on subsequent changes to these files.
3. PSOC6: Generate Cypress PSoC 6 target code (split from commit "Add Cypress PSoC 6 targets")
    - Would it be possible to provide an overview of how to recreate this in the commit message?
4. PSOC6: Integrate Cypress PSoC 6 target code (split from commit "Add Cypress PSoC 6 targets")
5. PSOC6: Update Cypress PSoC 6 code (split from commit "PSOC6: add Cypress target to mbedtls")
    - Changes here look like bug fixes unrelated to mbedtls. Can you confirm?
6. PSOC6: add Cypress target to mbedtls
7. PSoC6: add implementation of BLE HCI driver for CYW4343X/CYW43012 (unchanged)
8. Add WICED binary drivers for Cypress PSoC 6 targets (unchanged)

vmedcy committed:
PSOC6: reuse FUTURE_SEQUANA porting layer

Copy the porting layer from TARGET_PSOC6_FUTURE to TARGET_PSOC6.
This commit is intended to make the history and changes applied easier
to follow.
ipcpipe_transport.c, ipcpipe_transport.h, rpc_api.h, rpc_defs.h
are excluded (not used by Cypress port).
PeripheralNames.h is moved to BSP layer introduced in subsequent
commits (the peripheral names and count are board-specific).

vmedcy committed:
PSOC6: add the latest Cypress PDL sources

vmedcy committed:
PSOC6: update the HAL implementation to latest PDL

Integrate the latest fixes to the PSoC 6 HAL.

vmedcy committed:
PSOC6: add CRC and TRNG APIs

Exclude the labels from inherited FUTURE_SEQUANA targets.

vmedcy committed:
PSOC6: add BSPs for Cypress kits

vmedcy committed:
Add prebuilt CM0+ images for Cypress kits

vmedcy committed:
PSOC6: add BSP generated sources for Cypress kits

Code generated for pioneer kits:
* CY8CKIT-062-4343W
* CY8CKIT-062-BLE
* CY8CKIT-WIFI-BT
Prototyping boards:
* CY8CPROTO-062-4343W

The source is generated with ModusToolbox Device Configurator.
The origin design.modus files used to produce the GeneratedSource
will be submitted in the consequent pull requests.

vmedcy commented:
@c1728p9: thank you for reviewing.
> Would it be possible to update the history of this PR to make it easier to review?

Absolutely, I admit the PR is rather hard to review in the web interface. I mishandled the previous interactive rebase (squashed my updates with wrong base commit). Thanks a lot for you work on pr9481_history, indeed it is much easier to review this PR if all Cypress HAL fixes and GeneratedSource updates are separated. I rebased per your suggestions and provided more context in the commit messages. Now the commit history should be coherent.

I also included the latest HAL updates from the PDL team. I guess you are mostly focused on this commit: https://github.com/ARMmbed/mbed-os/pull/9481/commits/94c80819e0a2c6d98aaa75fb243fe8b6835417b5. @lrusinowicz: some Cypress HAL updates can also be applicable to FUTURE targets, kindly take a look.

vmedcy commented:
@cmonr: CapSense and WICED binary drivers were recompiled with IAR 7.8; we temporary switched our internal testing to use this old version (unfortunately there is incompatibility at C++ binary level). Whenever Mbed OS decides to switch to IAR 8, Cypress is ready to provide the new binaries.

There is also a great progress on having the greentea test suite pass. Latest result is:
**mbedgt: test case results: 1 FAIL / 618 OK / 1 ERROR**
Failing test cases: tests-mbed_hal-sleep and features-feature_ble-targets-target_cordio-tests-cordio_hci-transport (timeout).
Will attach the complete test logs once the remaining issues are addressed by the team, I guess there is no much point to provide intermediate results.

c1728p9 commented on targets/TARGET_Cypress/TARGET_PSOC6/analogin_api.c:
```diff
@@ -126,18 +131,22 @@ void analogin_init(analogin_t *obj, PinName pin)
             error("ANALOG IN pin reservation conflict.");
         }
-        obj->base = (SAR_Type*)CY_PERIPHERAL_BASE(sar);
+		
```
Nitpick - trailing whitespace here and in some other places in this commit

vmedcy replied:
@c1728p9: applied astyle to the HAL implementation.

c1728p9 commented on targets/TARGET_Cypress/TARGET_PSOC6/analogin_api.c:
```diff
@@ -126,18 +131,22 @@ void analogin_init(analogin_t *obj, PinName pin)
         }
-        obj->base = (SAR_Type*)CY_PERIPHERAL_BASE(sar);
+		
+        /* Initialize object */
```
Nitpick - a tab is used rather than 4 spaces here and in some other places in this commit

vmedcy replied:
@c1728p9: applied astyle to the HAL implementation.

c1728p9 reviewed (approved):
Hi @vmedcy thanks for the quick update. The HAL changes,  94c80819e0a2c6d98aaa75fb243fe8b6835417b5, look good to me. I left a couple of cosmetic nitpicks, but no actual problems. Thanks very much for updating the history. It made this PR very straight forward to review.

SeppoTakalo commented:
@KariHaapalehto Please review from IP-Core team point of view.

0xc0170 commented on features/FEATURE_BLE/targets/TARGET_Cypress/TARGET_CYW43012/HCIDriver.cpp:
```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright (c) 2018 ARM Limited
+ *
```
Can you add SPDX identifier (look at platform code for instance) here or to any new file? (follows our license guidelines)

vmedcy replied:
The copyright headers are defined by our legal department.
I think the SPDX identifier can be added with consecutive pull requests, once approved.
Please confirm this is not gating the initial PR.

0xc0170 replied:
> The copyright headers are defined by our legal department.

Should not be a blocker but our license guideline states that SPDX should be present. Please check with your legal (platform/driver code has it for instance).

vmedcy replied:
The SPDX identifier is already approved internally, will include with the next rebase (plan to get complete greentea test suite working later this week and submit the update).

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/pwmout_api.c:
```diff
@@ -156,59 +156,78 @@ void pwmout_init(pwmout_t *obj, PinName pin)
+            Cy_TCPWM_PWM_Disable(obj->base, obj->counter_id);
+            Cy_TCPWM_PWM_DeInit (obj->base, obj->counter_id, &pwm_config);
+        }
+        else
```
Not requirement at the moment, but I can see `{` starts on the new line here , not on thel ine 198 or others. If you can run astyle on HAL implementation, would be good to align the code

0xc0170 replied:
If run with astyle, it should address also above review about tabs/space alignment

vmedcy replied:
Applied astyle to the HAL implementation, now the style is aligned with mbed-os: https://github.com/ARMmbed/mbed-os/pull/9481/commits/e4e5f6538e8e2771de4dcf2f3e6255b14ede043c.
Note: PDL is using different code style (to match the company guidelines), no plan to reformat peripheral driver sources and device headers.
I plan to apply astyle to feature/mbedtls and feature/FEATURE_BLE targets on consecutive pull requests.

0xc0170 replied:
> Note: PDL is using different code style (to match the company guidelines), no plan to reformat peripheral driver sources and device headers.

Yes, please don't. Use astyleignore (currently ignored all in targets). We would like to allow HAL astyle checking in the future (no 3rd party drivers like PDL)

vmedcy replied:
Understood. Will notify mbedtls and BLE developers to apply mbed astyle to new Cypress code outside of /targets, and add astyle hook to internal CI to detect issues in our HAL implementation.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/trng_api.c:
```diff
@@ -0,0 +1,75 @@
+
+#define MAX_TRNG_BIT_SIZE         (32UL)
+
+void trng_init(trng_t *obj)
```
@ARMmbed/mbed-os-crypto Please review trng implementation

pan- commented:
@vmedcy Thanks for your answers. It raises other questions: 
* What should be the integration order between #9481 #9466 and #9509 ? If none of them work without the other; then we got an integration problem  (CC: @0xc0170 ). 
* What's missing besides `KitProg3` if I want to run `mbed test` (including flashing binaries) locally ?   
* Could you summarise integration plans of BLE on `CY8CKIT_062_BLE` ?

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/aes_alt.c:
```diff
@@ -0,0 +1,819 @@
+    memcpy(ctx->aes_buffers.keyInv, ctx->aes_state.invKey, CY_CRYPTO_AES_256_KEY_SIZE);
+
+    ctx->aes_state.buffers = (uint32_t *) &ctx->aes_buffers;
+    ctx->aes_state.key     = (uint8_t *) (ctx->aes_buffers.key);
```
Where are these pointers pointing before these assignments?

OlegKapshii replied:
(see 53d44d2 and a91822d merges)
SHA ALT - has been added there.
AES pointers - Driver for one of the HW Crypto block versions by default uses its internal memory and sets all needed pointers to it during aes driver init. In order to use external memory from context, we need to reassign the pointers

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/ecp_alt.h:
```diff
@@ -0,0 +1,233 @@
+}
+mbedtls_ecp_group;
+
+#if defined(MBEDTLS_ECP_RESTARTABLE)
```
This part can be removed, the MBEDTLS_ECP_RESTARTABLE feature does not work with alternative implementations.

OlegKapshii replied:
Will be cleaned in a next version of ecp_alt

kotkcy replied:
Done in 53d44d2

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/ecp_alt.c:
This part can be removed, the MBEDTLS_ECP_RESTARTABLE feature does not work with alternative implementations.

OlegKapshii replied:
Will be cleaned in a next version of ecp_alt

kotkcy replied:
Done in 53d44d2

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/ecp_alt.c:
We are using SECP256R1 arithmetic whatever the requested group is. This does not seem right.

OlegKapshii replied:
Fixed - see 53d44d2 and a91822d merges

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/ecp_alt.c:
With this we convert the point  at infinity to the origin unconditionally, this does not seem right. How does `Cy_Crypto_Core_EC_NistP_PointMultiplication()` represent the point at infinity? (I am confused, because at a superficial look it seems to use affine coordinates.)

OlegKapshii replied:
Will response later, need some internal discussion.

kotkcy replied:
Our crypto driver returns result in normalized Jacobian coordinates.
Z coordinate equal 1 in this case.

yanesca replied:
This means that if the result is the point (1,1,1) in normalized Jacobian coordinates then `Cy_Crypto_Core_EC_NistP_PointMultiplication()` returns (1,1). What does the function do if the result is the point at infinity ((1,1,0) in Jacobian coordinates)?

kotkcy replied:
Cy_Crypto_Core_EC_NistP_PointMultiplication() converts a point from Affine to Jacobian coordinates (Z=1) before an operations.
The result of operation converts from Jacobian projective coordinates to affine coordinates by Cy_Crypto_Core_JacobianInvTransform() before return:

/*******************************************************************************
* Function Name: Cy_Crypto_Core_JacobianInvTransform
****************************************************************************//**
*
* Transformation from Jacobian projective coordinates to affine coordinates in GF(VR_P).
* (s_x, s_y, s_z) -> (p_x, p_y), where p_x = s_x/s_z^2, p_y = s_y/s_z^3
********************************************************************************

What we should do to make a proper result data?

yanesca replied:
You can pass parameters to `Cy_Crypto_Core_EC_NistP_PointMultiplication()` such that the result is (1,1,0) in Jacobian coordinates. This point does not have a representation in affine coordinates (you would have to divide by zero in the conversion).

Since `Cy_Crypto_Core_EC_NistP_PointMultiplication()` represents the result in affine coordinates, the only sensible thing to do that comes to mind is to return an error. Then the caller could detect this error and set the point `R` to (1,1,0) manually .

pan- commented on features/FEATURE_BLE/targets/TARGET_Cypress/TARGET_CYW4343X/HCIDriver.cpp:
```diff
@@ -24,6 +24,7 @@
 #include <stdbool.h>
 #include "hci_mbed_os_adaptation.h"
 #include "H4TransportDriver.h"
+#include "cycfg_pins.h"
```
That file was used by [USI targets](https://github.com/ARMmbed/mbed-os/blob/ddc567d0d1fede0c18425689e48ae117a7a040f9/targets/targets.json#L2488-L2541) and is now inaccessible to these targets due to the `TARGET_Cypress` folder (these should be capital AFAIK). 

Please make sure the integration doesn't break existing targets.

vmedcy replied:
The TARGET_CYW4343X directory with HCI driver for targets inherited from USI_WM_BN_BM_22 was moved under TARGET_STM: https://github.com/ARMmbed/mbed-os/pull/9481/commits/22ed5e3d96f24c978bd4b7c446e3cb78ba1fecb0
I checked there is zero functional impact on MTB_USI_WM_BN_BM_22 and MTB_ADV_WISE_1530, thanks for noticing that before the CI.

features/FEATURE_BLE/targets now contains only vendor subdirectories. The TARGET_Cypress folder name is aligned with targets/TARGET_Cypress and features/mbedtls/targets/TARGET_Cypress

yanesca commented on features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/mbedtls_device.h:
```diff
@@ -0,0 +1,34 @@
+#define MBEDTLS_AES_ALT
+
+#define MBEDTLS_ECP_ALT
+#define MBEDTLS_ECP_DP_SECP192R1_ENABLED
```
These shouldn't be set by the driver, setting them should be left to the application.

OlegKapshii replied:
We have added only supported curves. May be it should be removed. Will discuss it.

kotkcy replied:
Addressed in dcfcbfc, macros for supported curves are commented out.

0xc0170 commented:
> What should be the integration order between #9481 #9466 and #9509 ? If none of them work without the other; then we got an integration problem (CC: @0xc0170 ).

The other two were integrated, this one should be fine. No conflicts at the moment. @vmedcy correct?

NirSonnenschein commented:
running CI while we wait for reviews to flush out any issues early.

SeppoTakalo reviewed (approved):
Approved from Mbed OS IP networking team point of view.

simonbutcher commented:
@cmonr - I don't see anything for the TLS team to review here. I think it's just down to @mbed-os-crypto / @Patater to review it.

vmedcy committed:
CYW4343X: move HCI driver to TARGET_STM

TARGET_CYW4343X Bluetooth HCI driver is specific to STM32 targets
derived from USI_WM_BN_BM_22. Move the driver implementation to
TARGET_STM folder to not interfere with Cypress implementation at
TARGET_Cypress/TARGET_CYW43XXX/HCIDriver.cpp that is enabled for
PSoC 6 targets which also include the CYW4343X label.

vmedcy committed:
PSoC6: add implementation of BLE HCI driver for CYW4343W/CYW43012

Add Cypress HCI driver implementation in TARGET_Cypress directory.

Update targets.json to enable CORDIO stack for Cypress PSoC 6 boards
with CYW43XXX radios with compatible HCI driver implementation:
CYW4343W and CYW43012.

vmedcy committed:
Add WICED binary drivers for Cypress PSoC 6 targets

Add prebuilt WICED libraries for the PSoC 6 boards with WICED radios:
* TARGET_CY8CKIT_062_4343W
* TARGET_CY8CKIT_062_WIFI_BT
* TARGET_CY8CPROTO_062_4343W
* TARGET_CYW943012P6EVB_01

Add required extra labels to the PSoC 6 targets in targets.json:
(PSOC6_01/02 labels are used during the library compilation).

vmedcy commented:
@pan-:
>     What should be the integration order between #9481 #9466 and #9509 ?

Everything except this PR is already merged to mainline (thanks @cmonr). I just rebased the PR branch against the latest master so you can fetch directly from pull/9481/head:add_cypress_kits

>     What's missing besides `KitProg3` if I want to run `mbed test` (including flashing binaries) locally ?

Make sure to update mbed-ls to 1.7.1 which was just released: https://github.com/ARMmbed/mbed-os-tools/releases/tag/v0.0.2 (thanks @bridadan).
We are still figuring out how to deliver the updated KitProg3.1 FW compatible with DAPLink.

>     Could you summarise integration plans of BLE on `CY8CKIT_062_BLE` ?

Porting of PSoC6 on-chip BLE is out of scope of initial release. The work is planned for the next release.

yanesca commented on targets/TARGET_Cypress/TARGET_PSOC6/trng_api.c:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    return (ret);
```
Here we are returning and leaving the last 4 byte of the TRNG output in memory, we should erase it before we return.

OlegKapshii replied:
Just to clarify that our understanding is correct - do you mean, that "random = 0" has to be added before return?

kotkcy replied:
Addressed in dcfcbfc. Random temporary buffer is cleared before exit from API.

yanesca replied:
@OlegKapshii Yes, that is what I meant!
@kotkcy thank you for the fix!

yanesca reviewed (changes requested):
As usual I have only reviewed the integration code and not the driver. I left my feedback and questions in the review comments.

I have a general question too: There seems to be a SHA accelerator as well, but it is not integrated with Mbed TLS. Is this intentional?

cmonr commented:
@sbutcher-arm I was under the impresion that you could review this commit: https://github.com/ARMmbed/mbed-os/pull/9481/commits/889d188375d7b0e3197ec65b60bbddd404290e03

Is that not the case?

cmonr commented:
>>   Could you summarise integration plans of BLE on `CY8CKIT_062_BLE` ?

> Porting of PSoC6 on-chip BLE is out of scope of initial release. The work is planned for the next release.

@vmedcy Making sure I understand, what does "next release" mean in this context? I thought this PR had support for BLE against the target.

vmedcy commented:
@cmonr: CY8CKIT_062_BLE has on-chip BLE subsystem. Other Cypress targets have either CYW4343W or CYW43012 Wi-Fi + Bluetooth combo. This pull request only provides BLE implementation for CYW43XXX radios. It doesn't enable BLESS support on CY8CKIT_062_BLE, such support will be submitted as a separate pull request targeting some of the future mbed-os releases. The work is not yet done.

deepikabhavnani reviewed (changes requested):
Binaries are added to this PR https://github.com/ARMmbed/mbed-os/pull/9481/commits/69c04007fb6c20e7197d30c8d64d4f8a38e48500, will like to know if IAR binaries are compatible to IAR 7.x and IAR 8.x version?

cydriftcloud commented:
> Binaries are added to this PR [69c0400](https://github.com/ARMmbed/mbed-os/commit/69c04007fb6c20e7197d30c8d64d4f8a38e48500), will like to know if IAR binaries are compatible to IAR 7.x and IAR 8.x version?

It is compatible to IAR 7.x only, which is the default supported IAR version.

vmedcy commented:
@deepikabhavnani, @cydriftcloud: correct, it is not feasible to provide IAR binaries that are compatible with both IAR 7 and IAR8, due to the linkage issues like #7733 and [inconsistent wchar_t size](https://www.iar.com/support/tech-notes/general/library-built-with-7.xx--causes-warning-message-in-8.11/). Once feature-iar8 is merged to master, we can rebuild libwiced_drivers.a with IAR 8.32.
What it the expected date of such merge? Do you accept PRs to feature/iar8 branch to make the transition smoother (avoid need to apply #9524 for Cypress targets)?

deepikabhavnani commented:
> Once feature-iar8 is merged to master, we can rebuild libwiced_drivers.a with IAR 8.32.

Binaries should be part of feature branch before it is merged to master. PR will be merged for 5.12 release. Path forward will depend on this PR's readinness

1. If PR Is ready for merge in master before or during 5.12 release binaries should be added to both master (IAR 7.x) and feature branch (IAR 8.x)
2. If PR is not ready for 5.12 release, IAR 8.x binary can be added to master directly

cmonr commented:
@vmedcy This is unfortunately an awkward time for introducing new targets, because of the IAR 8 update. Once `feature-iar8` is merged into master, PRs will then run against IAR 8.

Afaik, this is the flow of steps that will need to happen. It assumes that this PR will be ready before `feature-iar8` is:
- This PR is ready for CI, and contains IAR7 binaries.
- This PR passes and is merged.
- `feature-iar8` is rebased to pull in the new target.
- A PR is opened against `feature-iar8` to update the PSOC targets' binaries.
- PR passes, is merged.
- `feature-iar8` is merged into master.

The tldr for this PR is that it should use IAR 7 binaries, and after this is merged and closed, a PR against `feature-iar8` can be opened.

@bulislaw Fyi

pan- reviewed (approved):
I ran our ble test suite and the BLE driver provided works as expected with a CY43XXX controller (on a Wise1530 board). 
Integration is sufficient as well.

vmedcy commented:
@deepikabhavnani: we need to get this merged <= 5.12. Once this PR is merged, and feature-iar8 rebased, I can provide separate PR to feature-iar8 branch with IAR8 precompiled libraries.
@cmonr: the proposed plan is reasonable, is there any deadline for feature-iar8 being merged to master?

deepikabhavnani commented:
> the proposed plan is reasonable, is there any deadline for feature-iar8 being merged to master

Deadline for feature branch will near to code freeze for 5.12, PR for merge will be created and linked to #9206. Please track #9206, also it will be good to add IAR8.x version as soon as this PR has passes CI and is ready for merge

cmonr commented:
@vmedcy As I understand it, the last patch RC PR for 5.11 will be made on Feb 21st. Once that has passed and is merged, the CI will be updated once `feature-iar8` is opened for merge.

simonbutcher commented:
> I was under the impresion that you could review this commit: 889d188
> 
> Is that not the case?

@cmonr - That's the responsibility of my very capable colleague @Patater of @ARMmbed/mbed-os-crypto. It's not the task of the TLS team.

bulislaw commented:
Could be rebase feature-iar8 and ask for the updates there asap?

0xc0170 commented:
Besides iar8 coordination, we are waiting for crypto team to approve (they requested some changes). As this is quite an update, it would be really good to have  a note in the release notes for the release including this target update.

@vmedcy What should be mentioned there (the top decription provides useful info, is that OK to use it?).

@maclobdell @ashok-rao Can you review this PR by end of this week?

vmedcy commented:
@yanesca, @Patater: I rebased the PR to include all latest Cypress developments for crypto/mbedtls target. Please review 53d44d2 and a91822d.

vmedcy commented:
@0xc0170: travis-ci/licence_check failed due to blacklisted words in Cypress EULA (this is false positive: no GPL code in capsense middleware). I will figure out the proper license text with legal and update.

cmonr commented:
> Could be rebase feature-iar8 and ask for the updates there asap?

@bulislaw This PR needs to be merged first before any rebase of the feature branch will pull this in...

cmonr commented:
@0xc0170 https://github.com/ARMmbed/mbed-os/blob/372afd67a41be5053e6c04b11e474b1df3a5f4bc/targets/TARGET_Cypress/TARGET_PSOC6/psoc6mw/capsense/License.txt

It looks like `travis-ci/licence_check` found something, but it's unclear if this license should be in the codebase, or if it's ok to have.

bulislaw commented:
> Could be rebase feature-iar8 and ask for the updates there asap?

> @bulislaw This PR needs to be merged first before any rebase of the feature branch will pull this in...

Yes, I mean, we need update to platforms that will land as part of this PR...

cmonr commented:
CI started.
Testing since weekend load should be low, and this PR is quite involved

vmedcy commented:
@0xc0170 :
> What should be mentioned there (the top decription provides useful info, is that OK to use it?).

Yes, the top description of this PR (first half of it) is OK to include in the release notes.

vmedcy commented:
@cmonr, @0xc0170: 
> It looks like travis-ci/licence_check found something

This is now addressed with c77614b, travis CI passing. There are no licensing issues in psoc6mw/capsense.

vmedcy commented:
Good news: the team got the complete general greentea test suite passing.
Attached logs for CY8CKIT_062_BLE, CY8CKIT_062_WIFI_BT, CY8CPROTO_062_4343W, CYW943012P6EVB_01 with supported versions of GCC_ARM, ARM and IAR. No logs for CY8CKIT-062-4343W: kit is not yet plugged to our CI systems.

Attached mbedgt logs:
[CY8CKIT_062_BLE__ARM.log](https://github.com/ARMmbed/mbed-os/files/2824070/CY8CKIT_062_BLE__ARM.log)
[CY8CKIT_062_BLE__GCC_ARM.log](https://github.com/ARMmbed/mbed-os/files/2824071/CY8CKIT_062_BLE__GCC_ARM.log)
[CY8CKIT_062_BLE__IAR.log](https://github.com/ARMmbed/mbed-os/files/2824072/CY8CKIT_062_BLE__IAR.log)
[CY8CKIT_062_WIFI_BT__ARM.log](https://github.com/ARMmbed/mbed-os/files/2824073/CY8CKIT_062_WIFI_BT__ARM.log)
[CY8CKIT_062_WIFI_BT__GCC_ARM.log](https://github.com/ARMmbed/mbed-os/files/2824074/CY8CKIT_062_WIFI_BT__GCC_ARM.log)
[CY8CKIT_062_WIFI_BT__IAR.log](https://github.com/ARMmbed/mbed-os/files/2824075/CY8CKIT_062_WIFI_BT__IAR.log)
[CY8CPROTO_062_4343W__ARM.log](https://github.com/ARMmbed/mbed-os/files/2824076/CY8CPROTO_062_4343W__ARM.log)
[CY8CPROTO_062_4343W__GCC_ARM.log](https://github.com/ARMmbed/mbed-os/files/2824077/CY8CPROTO_062_4343W__GCC_ARM.log)
[CY8CPROTO_062_4343W__IAR.log](https://github.com/ARMmbed/mbed-os/files/2824078/CY8CPROTO_062_4343W__IAR.log)
[CYW943012P6EVB_01__ARM.log](https://github.com/ARMmbed/mbed-os/files/2824079/CYW943012P6EVB_01__ARM.log)
[CYW943012P6EVB_01__GCC_ARM.log](https://github.com/ARMmbed/mbed-os/files/2824080/CYW943012P6EVB_01__GCC_ARM.log)
[CYW943012P6EVB_01__IAR.log](https://github.com/ARMmbed/mbed-os/files/2824081/CYW943012P6EVB_01__IAR.log)

johndebudge commented:
Hello,
The Cypress internal software license review of this PR content is running a few days late.  Please do not accept this merge request until Cypress gives approval.  We expect approval by close of business Monday 2019-02-04.
Thank you,
-Cypress

vmedcy committed:
PSOC6: add network-default-interface-type to Cypress boards

Set the default network interface type as WIFI for Cypress
targets with on-board WICED Wi-Fi module.

maclobdell reviewed (approved):
Looks good to me.  I will continue to assist through the Mbed Enabled process and getting boards into CI.

0xc0170 commented:
Thanks @johndebudge for the info, we are waiting for your approval.

0xc0170 commented:
The files in mbed-os/targets/TARGET_Cypress/TARGET_PSOC6/psoc6mw/capsense/ are missing license headers . 
I can see there License.txt is not any of the listed permissive licenses we specify in the license guide. Please review

0xc0170 reviewed (changes requested):
Capsense license needs to be reviewed. Recommended to use Apache 2.0 as other parts of this PR

vmedcy committed:
Move FUTURE_SEQUANA port to TARGET_PSOC6_FUTURE

The targets/TARGET_Cypress/TARGET_PSOC6 is dedicated to the mbed-os HAL
and PSoC 6 MCU targets developed by Cypress Semiconductor. Move the
existing port developed by Future Electronics to TARGET_PSOC_FUTURE
and update the labels in targets.json appropriately.

vmedcy commented:
Hi,
I had to rebase this PR against latest master one more time to resolve merge conflict by picking https://github.com/ARMmbed/mbed-os/pull/9585.
I squashed the commits with PDL/BSP updates from last week with original PR to keep the git history clean.

@johndebudge comment still applies, resolution of license issues is pending.

NirSonnenschein commented:
@vmedcy t looks like you will need to do another rebase

vmedcy commented:
@NirSonnenschein: I will rebase immediately once the license issue is resolved

maclobdell commented:
There is an issue with the version of pyOCD specified by requirements.txt.  
the latest pyOCD release that supports these Cypress targets is 0.16, however, mbed-os/requirements.txt specifies the following
**pyocd>=0.14,<0.15**

I think this should be updated.  @cmonr @flit - can you confirm this is true?

Also in a separate but related issue, after I updated to pyOCD 0.16.0, I get this error, so mbed-host-tests needs to be updated too.  @bridadan 

pkg_resources.DistributionNotFound: The 'pyOCD==0.12.0' distribution was not found and is required by mbed-host-tests

vmedcy commented:
@NirSonnenschein: I rebased against latest master to pick caaaaff3e6d4c72fb155895d6e9d59564ee416f6.
@0xc0170: CapSense is removed from this patchset - all code in this PR is Apache 2.0 (hex files are under Permissive Binary License).

vmedcy commented:
@yanesca: thank you for the code review of mbedtls alternate functions. Cypress team faced issues integrating crypto driver with mbedtls, hence alternate implementations are now disabled in mbedtls_device.h (to enable stable networking using software implementations). Your feedback is now tracked internally, should not block this PR.

vmedcy commented:
> There is an issue with the version of pyOCD specified by requirements.txt.

@flit, @bridadan: can you please update requirements.txt to pull pyocd>=0.16.0 and mbed-ls>=1.7.1 (included by mbed-os-tools 0.0.2 which also includes mbed-host-tests 1.5.1)?

johndebudge commented:
> Hello,
> The Cypress internal software license review of this PR content is running a few days late. Please do not accept this merge request until Cypress gives approval. We expect approval by close of business Monday 2019-02-04.
> Thank you,
> -Cypress

The license review has been completed and the pull request can proceed.

bridadan commented:
> @flit, @bridadan: can you please update requirements.txt to pull pyocd>=0.16.0 and mbed-ls>=1.7.1 (included by mbed-os-tools 0.0.2 which also includes mbed-host-tests 1.5.1)?

I can if you feel like it'd be best to do that in a separate PR, otherwise I think you're fine making those changes in this PR.

vmedcy commented:
@bridadan: please update requirements.txt as a separate PR. I am not comfortable with pip version management. I also don't know how mbed-ls -> mbed-os-tools update is managed (as I understand mbed-greentea and mbed-host-tests are to be superseded by mbed-os-tools eventually).

0xc0170 commented:
Thanks for the update, I'll update my review soon

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6_FUTURE/sleep_api.c:
```diff
@@ -0,0 +1,42 @@
+void hal_deepsleep(void)
+{
+#if DEVICE_LPTICKER
+    if(CY_SYSPM_SUCCESS == Cy_SysPm_DeepSleep(CY_SYSPM_WAIT_FOR_INTERRUPT)) {
```
this file is just moved in this PR but why deepsleep only if lpticker is enabled? is there MCU that does not have lp ticker support thus no deepsleep?

vmedcy replied:
@0xc0170: all PSoC6 MCUs have LPTICKER, it is enabled in targets.json for MCU_PSOC6 base target.
The conditional check can be easily removed, without functional impact. For future releases, Cypress plans PSoC6 HAL cleanup.

0xc0170 commented:
@yanesca @lrusinowicz You requested changes while ago, can you review again? @vmedcy answered all the requests, we would like to start CI soon.

yanesca reviewed (changes requested):
@vmedcy Could you please move the addition of the crypto drivers in question (2e87bc9fcfc043316cff7eb2db7ed6fb7c30744a) to a separate PR for us to continue review and rework from? We can continue the review in the new PR, ensuring we don't add crypto drivers to Mbed OS before we've fixed any issues.

0xc0170 commented:
As requested change here, if this can be reworked asap, let us know. We will restart CI right away.

vmedcy commented:
@yanesca:
> @vmedcy Could you please move the addition of the crypto drivers in question ([2e87bc9](https://github.com/ARMmbed/mbed-os/commit/2e87bc9fcfc043316cff7eb2db7ed6fb7c30744a)) to a separate PR

I am afraid this is not possible with the existing timeline (moving mbedtls target to separate PR will exclude it from mbed-os-5.11.4/5.12.0). The current implementation of mbedtls alternate functions is working, but cannot be enabled by default in mbedtls_device.h due to the issues with simultaneous operation of SHA and AES crypto HW acceleration blocks, and incomplete curve support in ECP alternate implementation. There is still a desire to deliver the code in features/mbedtls/targets/TARGET_Cypress as part of this PR, to enable the HW acceleration for select customers who need this functionality in its current form.
For the future release, there is a plan to address all feedback on mbedtls implementation, enable all HW features in mbedtls_device.h and submit as separate PR for review. The work is tracked internally.

vmedcy commented:
@0xc0170: please confirm mbedtls target implementation can be merged as-is and improved later. Alternate functions are disabled in mbedtls_device.h but still accessible as part of mbed-os tree for reviewers and users that need it.

0xc0170 commented:
> I am afraid this is not possible with the existing timeline (moving mbedtls target to separate PR will exclude it from mbed-os-5.11.4/5.12.0

@vmedcy Please revert the commit referenced above (reverting adding crypto in this PR) as it requires further discussion and fixes. This will unblock the PR. Reverted commit will be added via a new pull request, crypto team and us will continue reviewing and making progress to enable it.

Let us know if you can revert it soon or this PR will be merged as it is, and we will revert in separate PR (will go to 5.11.4 also this one). We should be able to keep this within the timeline for 5.11.x.

vmedcy commented:
@0xc0170, @yanesca:
> Please revert the commit referenced above

This is done, removed 2e87bc9  from this PR.
Will create separate PR for mbedtls alternate implementation once the existing review feedback is addressed.

This PR is now ready for merge.

yanesca commented:
@vmedcy Thank you very much for your understanding!

Looks good to me!

adbridge commented:
CI started

SeppoTakalo commented:
@vmedcy Please note:

Seems that the NetworkInterface base class have been changed in https://github.com/ARMmbed/mbed-os/pull/9568, so these binaries needs to be rebuild against latest master.
Otherwise ARM and IAR builds are broken.

Sorry for the inconvenience.

SeppoTakalo commented:
Actually, I just realised.
This is targeting 5.11.4 and the #9568 is targeting 5.12. Therefore these binaries works against 5.11. branch, but do not work against master branch.

So I think we should merge this in as it is, do do a separate build of Wiced binaries from master. We still have time before 5.12 to fix that.

vmedcy commented:
@SeppoTakalo: thanks for letting us know. Since #9568 is already merged to master, is this pull request to be merged to a different branch to appear in 5.11.4, like mbed-os-5.11 or release-candidate? Is it safe to recompile libwiced_drivers.a now with the latest master and expect it to still work in 5.11.4?
@0xc0170: should I rebase this PR against master (with #9568), and ask team to recompile WICED drivers? Or this PR can be merged as-is? I see it already has "ready for merge" label - is it going to master in order to appear in 5.11.4?

adbridge commented:
@vmedcy All PRs got to master first and then we select specific ones to get cherry picked across to the release branch. So yes this would go to master first and then across to 5.11 ready for 5.11.4. So this should be fine in that release. As @SeppoTakalo said #9568 is targeting 5.12 so would not get ported across. So this can safely be merged in and go to the release as planned. You don't need to do anything else for now.

vmedcy commented:
@adbridge: thank you for explanation of release branching, and for PR approval.
@SeppoTakalo: I can confirm linkage errors of mbed-os-example-wifi with IAR/ARM in latest master due to changes in WiFiInterface vtable layout.
@deepikabhavnani: I guess you can now rebase feature-iar8 against master. We will provide libwiced_drivers.a built with IAR8 for TARGET_CY8CKIT_062_4343W, TARGET_CY8CKIT_062_WIFI_BT, TARGET_CY8CMOD_062_4343W, TARGET_CYW943012P6EVB_01.

SeppoTakalo commented:
@vmedcy Please hold before building new binaries for feature-iar8 branch (or other branches)

We can synch our changes with Multihoming branch, so that we need to rebuild once before 5.12 release. We need new binaries for master for ARM,GCC & IAR. Then we need new binaries for IAR8 and ARMC6.

There is now email discussion going between us & Cypress. Lets synch up off-line.

vmedcy commented:
@SeppoTakalo: this is understood, thanks. It makes sense to reduce the total number of rebuilds (to decrease dev/test effort).

deepikabhavnani commented:
@vmedcy - feature-iar8 is rebased, can add binaries after all changes for 5.12 are in sync.
